src/pretty.rs

```diff
@@ -0,0 +1,115 @@
+                            // non-special-case printing style
+                            _ => break,
+                        }
+                    }
+                }
+                (".", 2) => {
+                    let mut first = &args[0];
+                    let mut rest = vec![&args[1]]; // in reverse order for fast `push`ing
+                    while let Some((op, args)) = first.as_application() {
+                        match (op.display(sig).as_str(), args.len()) {
+                            (".", 2) => {
+                                first = &args[0];
+                                rest.push(&args[1]);
+                            }
+                            _ => break,
+                        }
+                    }
+                    rest.push(first);
+                    rest.reverse();
+                    let interior = rest.into_iter()
+                        .map(|x| x.pretty_inner(sig, false))
+                        .join(" ");
+                    return if spaces_allowed {
+                        interior
+                    } else {
+                        format!("({})", interior)
+                    };
+                }
+                ("CONS", 2) => {
+                    let mut items = vec![&args[0]];
+                    let mut cdr = &args[1];
+                    while let Some((op, args)) = cdr.as_application() {
+                        match (op.display(sig).as_str(), args.len()) {
+                            ("CONS", 2) => {
+                                items.push(&args[0]);
+                                cdr = &args[1];
+                            }
+                            ("NIL", 0) | ("NULL", 0) => {
+                                return format!(
+                                    "[{}]",
+                                    items
+                                        .into_iter()
+                                        .map(|item| item.pretty_inner(sig, true))
+                                        .join(", ")
+                                )
+                            }
+                            // list does not terminate with NIL, so we use the
+                            // non-special-case printing style
+                            _ => break,
+                        }
+                    }
+                }
+                _ => (),
+            }
+            let args_str = args.iter()
+                .map(|arg| arg.pretty_inner(sig, true))
+                .join(", ");
+            format!("{}({})", op_str, args_str)
+        } else {
+            self.display(sig)
+        }
+    }
+}
+impl Pretty for Context {
+    fn as_application(&self) -> Option<(Operator, &[Context])> {
+        match *self {
+            Context::Application { op, ref args } => Some((op, args)),
+            _ => None,
+        }
+    }
+    fn display(&self, sig: &Signature) -> String {
+        self.display(sig)
+    }
+}
```

Comment:
Just curious, is there any way to provide these as default implementations? Can you do some sort of `Self::Application...` type inference magic?

Reply:
I don't understand what you mean

Reply:
Ah I think I see what you mean. Unfortunately there's no way of representing particular enum variants as their own types, so an associated type for the trait in that manner can't be done. As for `Display`, this runs back into the issue that terms cannot be displayed without also having a relevant signature.